presto-main/src/main/java/com/facebook/presto/execution/buffer/SpoolingOutputBuffer.java

```diff
@@ -0,0 +1,721 @@
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Consumer;
+
+import static com.facebook.airlift.concurrent.Threads.daemonThreadsNamed;
+import static com.facebook.presto.execution.buffer.BufferResult.emptyResults;
+import static com.facebook.presto.execution.buffer.BufferState.FINISHED;
+import static com.facebook.presto.execution.buffer.BufferState.NO_MORE_PAGES;
+import static com.facebook.presto.execution.buffer.BufferState.OPEN;
+import static com.facebook.presto.spi.StandardErrorCode.STORAGE_INTERNAL_ERROR;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.base.Verify.verify;
+import static com.google.common.util.concurrent.Futures.immediateFailedFuture;
+import static com.google.common.util.concurrent.Futures.immediateFuture;
+import static com.google.common.util.concurrent.MoreExecutors.listeningDecorator;
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newCachedThreadPool;
+
+@SuppressWarnings("UnstableApiUsage")
+public class SpoolingOutputBuffer
+        implements OutputBuffer
+{
+
+    private final OutputBuffers outputBuffers;
+    private final StateMachine<BufferState> state;
+    private final String taskInstanceId;
+    private final TempDataOperationContext context;
+    private final TempStorage tempStorage;
+    private final DataSize threshold;
+    private final ListeningExecutorService executor;
+
+    private final AtomicLong totalBufferedBytes = new AtomicLong();
+    private final AtomicLong totalBufferedPages = new AtomicLong();
+    private final AtomicLong totalPagesAdded = new AtomicLong();
+    private final AtomicLong totalRowsAdded = new AtomicLong();
+
+    private final OutputBuffers.OutputBufferId outputBufferId = new OutputBuffers.OutputBufferId(0);
+    private final Closer closer = Closer.create();
+    private static ExecutorService coreExecutor = newCachedThreadPool(daemonThreadsNamed("spooling-outputbuffer-%s"));
+
+    @GuardedBy("this")
+    private final AtomicBoolean noMorePages = new AtomicBoolean();
+
+    @GuardedBy("this")
+    private final AtomicLong memorySequenceId = new AtomicLong();
+
+    @GuardedBy("this")
+    private final AtomicLong sequenceId = new AtomicLong();
+
+    @GuardedBy("this")
+    private final AtomicLong startPage = new AtomicLong();
+
+    @GuardedBy("this")
+    private final AtomicLong totalPagesRemaining = new AtomicLong();
+
+    @GuardedBy("this")
+    private final Queue<HandleInfo> handleInfoQueue = new ArrayDeque<>();
```

Comment:
I don't think we need queue semantics.  We can convert these to lists and simplify their use.

Reply:
Done